keps/sig-api-machinery/2839-in-use-protection/README.md

```diff
@@ -0,0 +1,612 @@
+- Protect `ConfigMap` while being used by `Pod`
+- Protect `StorageClass` while being used by `PersistentVolume`
+- Protect `VolumeSnapshot` while being used by `VolumeSnapshotContent`
+- Protect `VolumeSnapshotClass` while being used by `VolumeSnapshotContent`
+- Protect `User`, `Group`, `ServiceAccount` and `Role` from `RoleBinding`
+- Protect `User`, `Group`, `ServiceAccount` and `ClusterRole` from `ClusterRoleBinding`
+- User defined reference
+
+### Goals
+
+- Provide a generic mechanism that prevent X from being deleted when Y exists
+
+### Non-Goals
+
+- Provide a specific mechanism to decide which particular objects should be prevented from being deleted when other particular objects exist
+
+## Proposal
+
+A new controller to protect in-use objects from being deleted is introduced.
+This controller doesn't have special knowledge on which objects should be protected from deletion for which other objects, instead it gets such dependency information from objects that users or controllers set and protects these objects as defined there.
+
+### User Stories (Optional)
+
+#### Story 1
+
+Another controller to protect Secret from deletion watches all Secrets and their potential user objects, like `Pod`, `PersistentVolumes`, and `VolumeSnapshotContent`. It updates objects' dependency information and ask this in-use protection controller to protect the used objects while using objects exist.
+
+#### Story 2
+
+A user sets an object X's dependency information for a particular object Y to be protected by this in-user protection controller. For example, when a user knows that `Service` X shouldn't be erroneously deleted while `Deployment` Y exists for his/her particular use case, the user can set such a dependency.
+
+### Notes/Constraints/Caveats (Optional)
+
+- Cross namespace dependency isn't allowed by [garbage collector](https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#owners-and-dependents), however, cross namespace dependency will be required for in-use protection controller by the Secret protection use case above,
+- Cyclic dependency can be set by user, which potentially makes dead lock,
+- In some cases, users really wan't to delete the objects that are considered to be protected in general.
+For example, immutable Secrets might need to be deleted while it is used by a Pod to update the value.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+- Malicious users can add dependencies to objects that they don't have right to delete and can be prevent such objects from being deleted.
+
+## Design Details
```

Comment:
This design is very complicated, and I don't think it's necessary. I currently prefer a "lien" system, which is like finalizers except it prevents *beginning* the deletion process rather than just finishing it. This system is easier to implement and more general (works for conditions other than "object exists").

Reply:
@lavalamp 

Thank you for your comment.

> I currently prefer a "lien" system, which is like finalizers except it prevents beginning the deletion process rather than just finishing it. 

Yes. It is how the current pv-protection and pvc-protection works. So, if the secret-protection is allowed to be implemented in the same way, it will solve my concern.

The background of this KEP is to consider when and to which the finalizer be added (Note that pv-protection and pvc-protection adds their finalizers to all the PVs and PVCs on createion by using admission hook, regardless of whether it is actually used).

As commented [here](https://github.com/kubernetes/enhancements/pull/2640#discussion_r632691232) and [here](https://github.com/kubernetes/enhancements/pull/2640#discussion_r632692560), it is asked to add the finalizer to a resource only when the resource is actually used. I understand that this is for minimizing the impact of breaking the current behavior as commented [here](https://github.com/kubernetes/enhancements/pull/2640#discussion_r632695296), and it makes sense for this purpose.

However, by doing so, there will be a chance that the resource is deleted before the finalizer is added even when it is used.

Garbage collector is taking the approach to add finalizer only when there are owned resources, and it has been working well for a long time. Therefore, I created this KEP in the hope that this approach solves this issue (Also, the codes are rather big for each controller to implement, so I'm proposing this as a generic mechanism). Although, according to [this comment](https://github.com/kubernetes/enhancements/pull/2840#discussion_r677519448), this approach alone won't solve this issue ...

Any ideas or suggestions?

Reply:
> Yes. It is how the current pv-protection and pvc-protection works.

This is confusing, I'm not sure what you meant; we do not have liens right now. And finalizers don't prevent deletion from beginning.

> add the finalizer to a resource only when the resource is actually used

In a distributed, concurrent database, you have to do a little dance: 1. add the lien; 2. create your item; 3. on failure or on deletion of item, remove the lien.

I maintain that some variation of this dance happens no matter what, and it's best if it's explicit and the client controls it.

(IOW, I consider the proposal recorded here right now to be very racy. GC is also racy in at least one case but there's not simple alternatives.)

Reply:
> This is confusing, I'm not sure what you meant; we do not have liens right now.

Oh, I now understand what you mean by lien system. Lien is a different concept from finalizer. Finalizer allows beginning the deletion process but prevents finishing it until the finalizer is removed. On the other hand, lien prevents beginning the deletion process itself until the lien is removed. 

I still don't understand what issues are solved by the difference. Could you explain it?
Also, are there plans to implement it?

Reply:
One of the points seems that the lien is added before the item is created, so there seems no room to erroneously be deleted before the item is found to be created. Then, how we guarantee that the lien is deleted properly and not deleted by others erroneously?

Reply:
Liens don't exist yet and there aren't currently plans to add them. Given your problem statement, I expected this KEP to be about them :)

Currently deletion is irrevocable, and finalizers just slow the process down. The missing feature is some way to prevent the deletion in the first place. The requirements include that multiple actors must be able to independently preserve an object for different reasons. Liens accomplish this. I don't know if @deads2k has heard of any simpler alternative. (The "lien" name is from @bgrant0607, btw)

> how we guarantee that the lien is deleted properly and not deleted by others erroneously?

Well, I expect a KEP about liens to work that out. :) 

Also, what to do if a namespace is deleted but one of the objects in it has a lien on it?

Reply:
Thank you very much for your explanation.
It sounds the real solution for this KEP's scope.

Then, I will wait for the lien to be implemented and go back to the original discussion on how to handle the secret specific protection until the lien is implemented.

Reply:
Just to be clear, no one is working on it, there's not even a KEP.

Reply:
I mean to work without lien like [pv-protection](https://github.com/kubernetes/enhancements/issues/499) and [pvc-protection](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/postpone-pvc-deletion-if-used-in-a-pod.md), and eventually replace it with using lien once it becomes available.